Index: nginx-1.14.0/src/core/nginx.h
===================================================================
--- nginx-1.14.0.orig/src/core/nginx.h
+++ nginx-1.14.0/src/core/nginx.h
@@ -11,7 +11,7 @@

 #define nginx_version      1014000
 #define NGINX_VERSION      "1.14.0"
-#define NGINX_VER          "nginx/" NGINX_VERSION
+#define NGINX_VER          "nginx/" NGINX_VERSION " (EasyEngine)"

 #ifdef NGX_BUILD
 #define NGINX_VER_BUILD    NGINX_VER " (" NGX_BUILD ")"
